### ISSUE TITLE
Feature Suggestion: Recursion warning

### ISSUE THREAD

**Issue by halohalospecial** [stil4m/elm-analyse#41]
![](https://pbs.twimg.com/media/C6Uq4SXXEAILxfw.jpg)

**Comment by andys8** [comment]
It would be especially intersting if it's not tail recursive.
But recursion in general is not something that has to be fixed, right? So why do we want to have a warning?

**Comment by halohalospecial** [comment]
@andys8 You're right, the title was misleading.  I was aiming for just an indicator like in the screenshot.  The project owner can freely close this issue if it's out of scope :)

**Comment by andys8** [comment]
How about suggesting this feature with a visible indicator to the creators of the IDE plugins? For example [VScode](https://github.com/Krzysztof-Cieslak/vscode-elm) or [Elmjutsu for Atom](https://atom.io/packages/elmjutsu)

**Comment by stil4m** [comment]
@andys8, the issue was created by the author of Elmjutsu :). 

As for now, I want the violations in elm-analyse to be kept to "things that you do not want to have in your project", instead of "context-dependent bad".
Having the first property will decrease the need for issues such as #92, and increases the use of elm-analyse in combination with CI.

So, if this will be implemented, is there a well-known algorithm to detect if it is not tail-recursion?
I'm not really familiar with it.

**Comment by andys8** [comment]
Haha, I didn't know that. So I basically advised @halohalospecial to do it on his own ^^ By the way, great plugin @halohalospecial :)

**Comment by halohalospecial** [comment]
My initial thought was to have indicators to remind the devs that there are calls that may cause infinite recursions (which will only appear in production).  They may not be aware that calls are recursive (e.g. function A calls function B, which calls function C, which calls function A), or they may have written the incorrect function names (which they expected to be non-recursive).  The compiler doesn't notify them of these scenarios.

I understand that this feature may be out of scope since it is context-dependent and not really a bad practice.  It will just tell the devs to be wary about certain parts of the code and make sure that they really want recursion in those parts.

I'll ask Elmjutsu's creator if he is willing to implement this :laughing:

**Comment by andys8** [comment]
When talking about infinite recursion: There is 
[bad-recursion.md](https://github.com/elm-lang/elm-compiler/blob/dev/hints/bad-recursion.md) in the `elm-compiler` repository where @evancz describes what the compiler is currently doing. Quite an interesting read, if you don't know it already.

**Comment by halohalospecial** [comment]
Thanks for the link, @andys8!
